File: Dsa-day16.py
from collections import Counter

class Solution:
    def minCost(self, basket1: list[int], basket2: list[int]) -> int:
        combined_freq = Counter(basket1)
        combined_freq.update(basket2)

        min_cost = float('inf')
        for x in basket1:
            min_cost = min(min_cost, x)
        for x in basket2:
            min_cost = min(min_cost, x)

        misplaced = []
        freq1 = Counter(basket1)

        for fruit, total_count in combined_freq.items():
            if total_count % 2 != 0:
                return -1
            
            target_count = total_count // 2
            count1 = freq1.get(fruit, 0)
            
            if count1 > target_count:
                for _ in range(count1 - target_count):
                    misplaced.append(fruit)
            elif count1 < target_count:
                for _ in range(target_count - count1):
                    misplaced.append(fruit)
        
        misplaced.sort()
        
        total_cost = 0
        k = len(misplaced) // 2
        
        for i in range(k):
            total_cost += min(misplaced[i], min_cost * 2)
            
        return total_cost

File: Dsa-day16.java
import java.util.*;

class Solution {
    public long minCost(int[] basket1, int[] basket2) {
        Map<Integer, Integer> combinedFreq = new HashMap<>();
        long min_cost = Long.MAX_VALUE;

        for (int x : basket1) {
            combinedFreq.put(x, combinedFreq.getOrDefault(x, 0) + 1);
            min_cost = Math.min(min_cost, x);
        }
        for (int x : basket2) {
            combinedFreq.put(x, combinedFreq.getOrDefault(x, 0) + 1);
            min_cost = Math.min(min_cost, x);
        }

        List<Integer> misplaced = new ArrayList<>();
        Map<Integer, Integer> freq1 = new HashMap<>();
        for (int x : basket1) {
            freq1.put(x, freq1.getOrDefault(x, 0) + 1);
        }

        for (Map.Entry<Integer, Integer> entry : combinedFreq.entrySet()) {
            int fruit = entry.getKey();
            int totalCount = entry.getValue();
            if (totalCount % 2 != 0) {
                return -1;
            }
            int targetCount = totalCount / 2;
            int count1 = freq1.getOrDefault(fruit, 0);

            if (count1 > targetCount) {
                for (int i = 0; i < count1 - targetCount; i++) {
                    misplaced.add(fruit);
                }
            } else if (count1 < targetCount) {
                for (int i = 0; i < targetCount - count1; i++) {
                    misplaced.add(fruit);
                }
            }
        }

        Collections.sort(misplaced);
        
        long totalCost = 0;
        int k = misplaced.size() / 2;
        
        for (int i = 0; i < k; i++) {
            totalCost += Math.min(misplaced.get(i), min_cost * 2);
        }
        
        return totalCost;
    }
}